docs/explanation/holistic-vs-delta-charms.md

```diff
@@ -1,66 +1,252 @@
+
+### Example
+
+```py
+def __init__(self, framework: ops.Framework):
+    super().__init__(framework)
+    self.workload = Workload()
+    self.framework.observe(self.on.install, self._on_install)
+    self.framework.observe(self.on.start, self._on_start)
+
+    hostname = socket.getfqdn()
+    self.foo_requirer = FooRequirer(self, "foo-relation", address=hostname)
+    self.framework.observe(
+        self.foo_requirer.on.data_available,
+        self._on_data_available,
+    )
+
+    self.bar_provider = BarProvider(self, "bar-relation")
+    self.framework.observe(
+        self.bar_provider.on.create_bar,
+        self._on_create_bar,
+    )
+
+def _on_install(self, event: ops.InstallEvent):
+    self.workload.install_binaries()
+
+def _on_start(self, event: ops.StartEvent):
+    self.workload.start_service()
+    # Peer relation is now usable
+    if self.unit.is_leader():
+        ...
+
+def _on_data_available(self, event: DataAvailableEvent):
+    # Update the workload with event's data
+    self.workload.reconfigure(some_key=event.some_value)
+
+def _on_create_bar(self, event: CreateBarEvent):
+    # Provision a `Bar` resource in the workload
+    self.workload.create_bar(event.some_field)
 ```
 
+Notice how the handler methods map to Juju and custom events directly.
+There's less boilerplate and flow control is predictable.
+
+### Developing your charm
+
+The charm code and tests can get messy when event interdependencies are accounted for:
+
+- creating a resource is only possible after the workload has been installed
+- the available data typically needs to be mixed with application's configuration
+- cases where something is inferred from the sequence of events (see {external+juju:ref}`Juju | Hook ordering <hook>`)
+
+A good rule of thumb is this: if you're starting to use `defer` in various places, consider whether it's time to rewrite the charm using the reconciler pattern.
+
+## Which approach to use?
+
+At a high level, simple workloads are served well by delta charms.
+Complex workloads are more robust if the reconciler pattern is followed.
+The reconciler pattern is especially suitable for mature, feature-rich charms that use several charm libraries.
+
+### Dedicated handlers for specific Juju events
+
+Some events are special, and are typically processed outside the reconciler pattern in holistic charms.
+In both styles of charms a dedicated event handler is used for:
 
-## When to use the holistic approach
+- unit lifecycle end events, because the goal of reconciliation is different
+- action event, as it is synchronous and the event object holds arguments and results
+- secret rotation event, where secret must be modified synchronously
+- secret remove event, because the revision that needs removal is not available otherwise
```

Comment:
```suggestion
- the `secret-remove` event, because the revision that needs removal is not available otherwise
```

Reply:
Sorry, I've confused secret-expired and secret-changed yet again.